File: app/src/main/java/com/dougkeen/bart/backend/TransitFeedFetchResult.java
package com.dougkeen.bart.backend;

import com.google.transit.realtime.GtfsRealtime;

/** Results from independently fetched trip-update and alert feeds. */
public final class TransitFeedFetchResult {
    private final GtfsRealtime.FeedMessage tripUpdates;
    private final GtfsRealtime.FeedMessage alerts;
    private final Exception tripUpdatesError;
    private final Exception alertsError;
    private final TransitFeedSnapshot completeSnapshot;

    public TransitFeedFetchResult(GtfsRealtime.FeedMessage tripUpdates,
                                  Exception tripUpdatesError,
                                  GtfsRealtime.FeedMessage alerts,
                                  Exception alertsError) {
        if (tripUpdates == null && tripUpdatesError == null) {
            throw new IllegalArgumentException(
                    "Trip updates need data or an error");
        }
        if (alerts == null && alertsError == null) {
            throw new IllegalArgumentException("Alerts need data or an error");
        }
        this.tripUpdates = tripUpdates;
        this.tripUpdatesError = tripUpdatesError;
        this.alerts = alerts;
        this.alertsError = alertsError;
        this.completeSnapshot = null;
    }

    public static TransitFeedFetchResult complete(TransitFeedSnapshot snapshot) {
        if (snapshot == null) {
            throw new NullPointerException("snapshot");
        }
        return new TransitFeedFetchResult(snapshot.getTripUpdates(), null,
                snapshot.getAlerts(), null, snapshot);
    }

    public static TransitFeedFetchResult failed(Exception exception) {
        if (exception == null) {
            throw new NullPointerException("exception");
        }
        return new TransitFeedFetchResult(null, exception, null, exception);
    }

    public GtfsRealtime.FeedMessage getTripUpdates() {
        return tripUpdates;
    }

    public GtfsRealtime.FeedMessage getAlerts() {
        return alerts;
    }

    public Exception getTripUpdatesError() {
        return tripUpdatesError;
    }

    public Exception getAlertsError() {
        return alertsError;
    }

    TransitFeedSnapshot getCompleteSnapshot() {
        return completeSnapshot;
    }

    private TransitFeedFetchResult(GtfsRealtime.FeedMessage tripUpdates,
                                   Exception tripUpdatesError,
                                   GtfsRealtime.FeedMessage alerts,
                                   Exception alertsError,
                                   TransitFeedSnapshot completeSnapshot) {
        this.tripUpdates = tripUpdates;
        this.tripUpdatesError = tripUpdatesError;
        this.alerts = alerts;
        this.alertsError = alertsError;
        this.completeSnapshot = completeSnapshot;
    }
}

File: app/src/main/kotlin/in/izyum/bart/backend/TransitFeedFetchResult.kt
package `in`.izyum.bart.backend

import com.google.transit.realtime.GtfsRealtime

/** Results from independently fetched trip-update and alert feeds. */
class TransitFeedFetchResult private constructor(
    val tripUpdates: GtfsRealtime.FeedMessage?,
    val tripUpdatesError: Exception?,
    val alerts: GtfsRealtime.FeedMessage?,
    val alertsError: Exception?,
    private val completeSnapshot: TransitFeedSnapshot?
) {
    constructor(
        tripUpdates: GtfsRealtime.FeedMessage?,
        tripUpdatesError: Exception?,
        alerts: GtfsRealtime.FeedMessage?,
        alertsError: Exception?
    ) : this(
        tripUpdates,
        tripUpdatesError,
        alerts,
        alertsError,
        null
    ) {
        require(tripUpdates != null || tripUpdatesError != null) {
            "Trip updates need data or an error"
        }
        require(alerts != null || alertsError != null) {
            "Alerts need data or an error"
        }
    }

    fun getCompleteSnapshot(): TransitFeedSnapshot? = completeSnapshot

    companion object {
        @JvmStatic
        fun complete(snapshot: TransitFeedSnapshot?): TransitFeedFetchResult {
            val value = requireNotNull(snapshot) { "snapshot" }
            return TransitFeedFetchResult(
                value.tripUpdates,
                null,
                value.alerts,
                null,
                value
            )
        }

        @JvmStatic
        fun failed(exception: Exception?): TransitFeedFetchResult {
            val value = requireNotNull(exception) { "exception" }
            return TransitFeedFetchResult(null, value, null, value)
        }
    }
}
